src/core/mesh/qgsmesheditor.cpp

```diff
@@ -1482,3 +1491,152 @@ void QgsMeshLayerUndoCommandAdvancedEditing::redo()
       mMeshEditor->applyEdit( edit );
   }
 }
+
+QgsMeshLayerUndoCommandAddVertexInFaceWithDelaunayRefinement::QgsMeshLayerUndoCommandAddVertexInFaceWithDelaunayRefinement(
+  QgsMeshEditor *meshEditor,
+  const QgsMeshVertex vertex,
+  double tolerance )
+  : QgsMeshLayerUndoCommandMeshEdit( meshEditor )
+  , mVertex( vertex )
+  , mTolerance( tolerance )
+{
+  setText( QObject::tr( "Add vertex inside face with Delaunay refinement" ) );
+}
+
+void QgsMeshLayerUndoCommandAddVertexInFaceWithDelaunayRefinement::redo()
+{
+  if ( !mVertex.isEmpty() )
+  {
+    QgsMeshEditor::Edit edit;
+
+    mMeshEditor->applyAddVertex( edit, mVertex, mTolerance );
+    mEdits.append( edit );
+
+    QList<std::pair<int, int>> sharedEdges = innerEdges( triangularFaces( facesContainingVertex( mMeshEditor->topologicalMesh().mesh()->vertexCount() - 1 ) ) );
+
+    for ( std::pair<int, int> edge : sharedEdges )
+    {
+      if ( mMeshEditor->edgeCanBeFlipped( edge.first, edge.second ) && !mMeshEditor->topologicalMesh().delaunayConditionForEdge( edge.first, edge.second ) )
+      {
+        mMeshEditor->applyFlipEdge( edit, edge.first, edge.second );
+        mEdits.append( edit );
+      }
+    }
+
+    mVertex = QgsMeshVertex();
+  }
+  else
+  {
+    for ( QgsMeshEditor::Edit &edit : mEdits )
+      mMeshEditor->applyEdit( edit );
+  }
+}
+
+QSet<int> QgsMeshLayerUndoCommandAddVertexInFaceWithDelaunayRefinement::facesContainingVertex( const int vertexId )
+{
+  QList<int> facesAroundVertex;
+
+  for ( int i = 0; i < mMeshEditor->topologicalMesh().mesh()->faceCount(); i++ )
+  {
+    QgsMeshFace face = mMeshEditor->topologicalMesh().mesh()->face( i );
+
+    if ( face.contains( vertexId ) )
+    {
+      facesAroundVertex.push_back( i );
+    }
+  }
```

Comment:
This is brute forcing the faces to find the ones around the vertex, and later on are brute forced again.
This can be expensive with large datasets.
Maybe this can be optimized by using `mMeshEditor->topologicalMesh().facesAroundVertex();`
I think both `facesContainingVertex()` and `triangularFaces()` methods could be replace by something like:
```cpp
    const int vIndex = mMeshEditor->topologicalMesh().mesh()->vertexCount() - 1;
    const QList<int> firstNeighborFaces = mMeshEditor->topologicalMesh().facesAroundVertex( vIndex );
    QSet<int> firstNeighborVertices;
    for ( int face : firstNeighborFaces )
    {
      const QgsMeshFace meshFace = mMeshEditor->topologicalMesh().mesh()->face( face );
      for ( int vertex : meshFace )
      {
        firstNeighborVertices.insert( vertex );
      }
    }

    QSet<int> secondNeighboringFaces;
    for ( int vertex : firstNeighborVertices )
    {
      const QList<int> faces = mMeshEditor->topologicalMesh().facesAroundVertex( vertex );
      for ( int face : faces )
      {
        if ( mMeshEditor->topologicalMesh().mesh()->face( face ).count() == 3 )
          secondNeighboringFaces.insert( face );
      }
    }
    const QList<std::pair<int, int>> sharedEdges = innerEdges( secondNeighboringFaces );
```